### ISSUE TITLE
[Feature] Use Directory.Packages.Props for single version principle instead of .nx-dotnetrc.json

### ISSUE THREAD

**Issue by AgentEnder** [nx-dotnet/nx-dotnet#562]
**Is your feature request related to a problem? Please describe.**
Currently, we use a field in .nx-dotnetrc.json for single version principle. There are some benefits to this:
- We control the file, its json, its easier to read / update than an xml file.
- Since its custom, .NET SDK version doesn't matter.

However, its non-standard. `dotnet` doesn't know about this file, and we have to track packages when they are added. We have a `sync` generator thats entire purpose is to keep up with this stuff, and its a really messy area in the codebase.

**Describe the solution you'd like**
.NET 7 introduces some capabilities for centralized package management. They use `Directory.Packages.props` for this.
https://devblogs.microsoft.com/dotnet/announcing-dotnet-7/#central-package-management

We already make use of other directory props files for adding tasks, so this makes sense for us.

**Additional context**
This feature is only supported in .NET 7+, so some thought will have to be given about supporting existing users of .NET 6 which is LTS. See lifecycle chart here: https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-and-net-core

**Comment by AgentEnder** [comment]
@tzuge do you have any thoughts on this one? I think it could simplify things quite nicely.

**Comment by tzuge** [comment]
I agree it would be nice to let the dotnet tool stack handle this behaviour. I'm curious about how it will work with the dotnet cli. If nx-dotnet still has to do special handling (e.g. xml manipulation in the props file), then that seems not as beneficial.

**Comment by AgentEnder** [comment]
At the end of the day, we will still have to do **_some_** manipulation in the props file most likely. I think itd still be worth it since it would move the manipulation outside of the .csproj, which is more likely to have some weird custom stuff

**Comment by bcallaghan-et** [comment]
Microsoft has actual documentation for [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management) that is slightly more informative than the announcement blog above. This page specifies that .NET SDK 6.0.300+ is compatible with the feature, not just .NET 7+.

**Comment by vzakanj** [comment]
I'm actually using this already and the way it currently works is CPM takes precedence - the package versions are added to the Directory.Packages.Props, the csproj files have nuget references without the version specified, and nx-dotnet.rc.json also contains nuget references without package versions, e.g.

```json
{
  "nugetPackages": {
    "FastEndpoints": "",
    "Humanizer": ""
  }
}
```

I'm not familiar with the internals of nx-dotnet yet, but the way I understand it now, there's no need for nx to manipulate the props files since dotnet add package already takes care of it.
So either detecting if CPM is used or using a configuration flag which would enable handing off nuget versioning completely to cpm would be a good solution to this?  (i.e. not using nx-dotnet.rc.json at all in that case)

**Comment by Tungsten78** [comment]
I can confirm `dotnet add {PROJECT.CSPROJ} package {NUGET_PACKAGE}` will auto-detect CPM - modifying `Directory.Packages.props` and the csproj automatically.

**Comment by AgentEnder** [comment]
Is there a dotnet command that would set up CPM? If so, I'd happily drop managing nuget versions.

**Comment by Tungsten78** [comment]
Not that I've found, but the steps are pretty concise:
1) Add Directory.Packages.props - I think it could start life as simply `<Project></Project>`
2) Add `<ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>` to Directory.Build.Props

I assume from there dotnet cli would just align with CPM (dotnet add, dotnet new).

I do wonder tho, this kinda eliminates any need for the nuget generators in nx-dotnet.  no?

This [extension ](https://marketplace.visualstudio.com/items?itemName=patcx.vscode-nuget-gallery) seems to work nicely in our nx repo.

**Comment by AgentEnder** [comment]
Yeah, this would eliminate @nx-dotnet/core:nuget-reference as well as @nx-dotnet/core:sync, and also the configuration flags for nuget packages entirely.

It would be a pretty large win 😄.

Would you be up for contributing the change? I haven't worked inside of a .NET repo in over a year personally, so I've not spent time learning how some of the newer stuff works.

**Comment by Tungsten78** [comment]
Boo,  just found that `dotnet new` is not resulting in a CPM compatible csproj - it adds the version attributes which result in build errors.  

I'll research this a bit more tomorrow.

**Comment by AgentEnder** [comment]
Sounds good - I'm working on some configuration changes right now anyways so it wouldn't be a great idea to start working on the changeset until I merge them to avoid conflicts.

**Comment by AgentEnder** [comment]
This is the mentioned PR w/ configuration changes: https://github.com/nx-dotnet/nx-dotnet/pull/787

**Comment by Tungsten78** [comment]
Confirmed, `dotnet new` is blind to CPM.

I think it's a minor inconvenience to manually remove `Version` attributes from the generated csproj files.
* **I believe this is outside the scope of nx-dotnet**  - @AgentEnder  agree?

To see this:
```
dotnet "new" "xunit" "--language" "C#" "--name" "NxTrends.Test" "--output" "libs/test"            
The template "xUnit Test Project" was created successfully.

Processing post-creation actions...
Restoring C:\Users\Chris Leigh\Source\repos\nx-trends\libs\test\NxTrends.Test.csproj:
  Determining projects to restore...
C:\Users\Chris Leigh\Source\repos\nx-trends\libs\test\NxTrends.Test.csproj : error NU1008: Projects that use central package version management should not define the version on the PackageReference items but on the PackageVersion items: Micr
osoft.NET.Test.Sdk;xunit;xunit.runner.visualstudio;coverlet.collector.
  Failed to restore C:\Users\Chris Leigh\Source\repos\nx-trends\libs\test\NxTrends.Test.csproj (in 79 ms).
Restore failed.
Post action failed.
Manual instructions: Run 'dotnet restore'
```

**Comment by AgentEnder** [comment]
I think the project generators would need to handle moving version specifiers to the Directory.Packages.Props file (warning if there was a mismatch), otherwise the behavior would result in this failing:

```
nx g app my-api --template webapi --language C#
nx build my-api
```

It should be fine to migrate all repos to CPM, we just need to take care to not break folks who werent really using the existing behavior.

**Comment by Tungsten78** [comment]
The presence of the packages file will force folks into CPM. Meaning they
will see the restore errors similar to what’s above.  They could of course
disable CPM it from the build.props.   But then versions have to be
specified on package references.  So it’s all or nothing.

Using single versioning beforehand meant using the nuget generator.  Mixing
was possible.

Not sure if the better route is to have user opt in at init. And have
generators detect if it’s enabled and act accordingly.  This leaves it as
an option but doesn’t force it.   This could look like no migration but a
generator to turn it on after the fact.

Thoughts?

**Comment by AgentEnder** [comment]
Using central package management works, and shouldn't cause an issue, this issue is just for adding some optimizations to the plugin for it

**Comment by bcallaghan-et** [comment]
@brinehart My team uses Nx in a large .NET repository with Central Package Management. The executors work just as well, though I haven't tested any of the generators. The one notable issue I've found is that changing any NuGet package in Directory.Packages.props results in a full rebuild of the entire repository (#951).

**Comment by Tungsten78** [comment]
@brinehart - I can also confirm, using CPM works fine.    

- Enable it manually [docs](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#enabling-central-package-management).    
- Of note:  the generators will trigger a restore error due to the version attribute in the csproj package reference (if I recall correct, only the test library), but this is called out in the error and easy to spot fix.
